// dojs/lib/redisSession.js
/**
 * Session类, 存入redis
 * @authors yutent (yutent@doui.cc)
 * @date    2016-03-14 16:08:57
 *
 */
"use strict";

let ioredis = require('ioredis')

class Session {

    constructor(){

        let conf = dojs.session.conf

        this.data = {}
        
        this.redis = new ioredis({
            host: conf.host || '127.0.0.1',
            port: conf.port || 6379,
            db: conf.db || 0
        })
    }


    createSsid(ssid){
        let ua = Request.headers('user-agent')
        ua = libs.SEC.md5(ua)
        if(!ssid || ssid.slice(-32) !== ua)
            ssid = 'sid:' + libs.SEC.md5(Date.now()).slice(0, 8) + ua

        //客户端有响应时, 即时更新cookie, 以防SESSION过期
        cookie('NODESSID', ssid, {maxAge: dojs.sessionTTL, httpOnly: true})
        this.ssid = ssid
        //设置session有效期
        this.redis.expire(ssid, dojs.session.ttl)
    }


    start(){
        this.createSsid(cookie('NODESSID'))
    }

    config(ssid){
        if(!ssid)
            return this.ssid
        else
           this.createSsid(ssid) 
    }


    //获取session字段值, 需要yeild指令
    get(key){
        //不传key时,直接返回全部字段
        if(!key){
            return fn => {
                this.redis.hgetall(this.ssid, fn)
            }
        }

        if(this.data.hasOwnProperty(key)){
            return fn => {
                fn(null, this.data[key])
            }
        }

        return fn => {
            this.redis.hget(this.ssid, key, (err, val) => {
                if(err || !val)
                    return fn(err, null)

                this.data[key] = val
                fn(err, val)

            })
        }
    }


    //设置session字段值
    set(key, val){
        this.data[key] = val
        this.redis.hset(this.ssid, key, val)
    }


    //删除单个字段
    unset(key){
        if(this.data.hasOwnProperty(key))
            delete this.data[key]

        this.redis.hdel(this.ssid, key)
    }

    //清除个人session
    clear(){
        this.data = {}
        this.redis.del(this.ssid)
    }

}


module.exports = new Session
