// dojs/lib/session.js
/**
 * 
 * @authors yutent (yutent@doui.cc)
 * @date    2016-03-15 16:01:38
 *
 */
"use strict";


class Session {

    constructor(){
        this.data = {}
    }

    createSsid(ssid){
        let ua = Request.headers('user-agent')
        ua = libs.SEC.md5(ua)
        if(!ssid || ssid.slice(-32) !== ua){
            this.data = {}
            ssid = 'sid:' + libs.SEC.md5(Date.now()).slice(0, 8) + ua
            cookie('NODESSID', ssid, {httpOnly: true, domain: dojs.session.domain})
        }
        this.ssid = ssid
        
        if(!this.data.hasOwnProperty(ssid) || this.data[ssid].expire < Date.now())
            this.data[ssid] = {}
        //设置session有效期
        this.data[ssid].expire = Date.now() + dojs.session.ttl * 1000
    }

    start(){
        this.createSsid(cookie('NODESSID'))
    }

    config(ssid){
        if(!ssid)
            return this.ssid
        else
           this.createSsid(ssid) 
    }


    //获取session字段值, 需要用yield命令,这是为了兼容redis写法
    get(key){
        return fn => {
            fn(null, key ? (this.data[this.ssid][key] || null) : this.data[this.ssid])
        }
        
    }


    //设置session字段值
    set(key, val){
        this.data[this.ssid][key] = val
    }


    //删除单个字段
    unset(key){
        if(this.data[this.ssid].hasOwnProperty(key))
            delete this.data[this.ssid][key]
    }

    //清除个人session
    clear(){
        delete this.data[this.ssid]
    }


}


module.exports = new Session
